[AddressLowering] Project at storage.

nate-chandler opened the pull request:
The pass rewrites opaque values in reverse post order of their definitions.  When an opaque value is rewritten, both its definition and its uses are rewritten.  When a def or a use is rewritten, in order to materialize an address with which to rewrite, projection instructions are created.  Previously, these projections were created at the site of the def or use.  Some of these projection instructions may be reused when rewriting later opaque values.

As a result, it's possible to have two opaque values `A` and `B` (which are rewritten in that order) such that rewriting a use of `A` which occurs "after" the def (or a use) of `B` creates a projection `P` which is then used by that "earlier" def (or use) of `B`:

    ```
    A =
    B =     // relies on P
    use B
    use A   // creates some projection P
    ```

When rewriting the def (or that use, respectively) of `B`, the projection which was created for the use of `A` will be reused.  And previously, the projection would be created at the use of A.  But that projection instruction was "after" the place where it is used when rewriting the def or use of `B`.  That's invalid!

To address this, rather than creating projections at the instruction being rewritten, instead create them "as early as possible".  The locations where they will be created are chosen so as to dominate all uses.

nate-chandler committed:
[AddressLowering] Gardening: Removed comment.

Per Andrew Trick, its author.

nate-chandler committed:
[AddressLowering] NFC: Added const variant.

Of existing function getProjectedStorage.

nate-chandler committed:
[AddressLowering] NFC: Add projection iterator.

Used the iterator to replace recursion in get*Storage functions with for
loops.

nate-chandler committed:
[AddressLowering] NFC: Add more dump functions.

nate-chandler committed:
[AddressLowering] NFC: Extracted function.

Promoted the functionality from createStackAllocation to a named member
AddressLoweringState::getLatestOpeningInst  so that it can be employed
elsewhere.

nate-chandler committed:
[AddressLowering] Place projections below types.

Projections may involve opened archetypes.  They must be dominated by
the instructions which open those archetypes.

When determining the earliest storage point for a value and the point at
which to insert its projections, look for such obstructions in the
projection chain.  The first one found is the earliest storage point.

nate-chandler commented:
@swift-ci please test

atrick commented on lib/SILOptimizer/Mandatory/AddressLowering.cpp:
```diff
@@ -1063,6 +1084,10 @@ void ValueStorageMap::recordComposingUseProjection(Operand *oper,
   }
+  // FIXME: FIXME_NOW: Something like this...
+  if (userValue->getType().isExistentialType()) {
+    storage.initializesEnum = true;
```
We should comment that `initializesEnum` is for the payload, not the enum. It maps to the instruction that initializes the payload, not the `enum` instruction.

nate-chandler replied:
Fixed.

atrick commented on lib/SILOptimizer/Mandatory/AddressLowering.cpp:
```diff
@@ -1712,11 +1763,41 @@ AddressMaterialization::materializeProjectionIntoUse(Operand *operand,
+
+SILInstruction *
+AddressLoweringState::getProjectionInsertionPoint(SILValue userValue) const {
+  if (!valueStorageMap.contains(userValue)) {
+    return userValue->getDefiningInsertionPoint();
+  }
+  if (!valueStorageMap.getStorage(userValue).isProjection()) {
+    return userValue->getDefiningInsertionPoint();
```
Hopefully these two special cases go away if moved inside AddressMaterialization. If not, they could be folded into a single helper.

nate-chandler replied:
Replaced both with defaulting `retval = userValue->getDefiningInsertionPoint();` before the loop.

atrick commented on lib/SILOptimizer/Mandatory/AddressLowering.cpp:
```diff
@@ -1712,11 +1763,41 @@ AddressMaterialization::materializeProjectionIntoUse(Operand *operand,
+  SILInstruction *retval = nullptr;
+  for (auto *pair : valueStorageMap.getProjections(userValue)) {
+    auto const &storage = pair->storage;
+    if (storage.storageAddress) {
+      // There's already an address.  Projections should be inserted after it.
+      if (auto *arg = dyn_cast<SILArgument>(storage.storageAddress)) {
+        // The address is a function argument.  Inserting before the first
+        // instruction will insert after it.
+        retval = &arg->getParent()->front();
+        break;
+      }
+      // The address is an instruction.  To insert after it, insertion must be
+      // before the next instruction.
+      auto *instruction = storage.storageAddress->getDefiningInstruction();
+      retval = instruction->getNextInstruction();
+      break;
+    }
```
This if-block does one simple task: return the next instruction after the storage address. It could be a helper on the storage address.

nate-chandler replied:
Fixed.  (ValueBase::getNextInstruction()`)

atrick reviewed (commented):
Great!

Let's be clear though that the broader intention is not for SIL to reuse address projections across blocks. I think that's a bad idea and hides bugs.

If you can get sinkAddressProjection to work, that would be great. Otherwise, may be you can add a TODO to document the intention.

nate-chandler committed:
[AddressLowering] Gardening: Detypo'd.

nate-chandler committed:
[AddressLowering] Gardening: Undefaulted argument.

That there was a default wasn't used anywhere.

nate-chandler committed:
[SIL] Promoted least-common ancestor to Dominance.

It's a general purpose helper and there will soon be another user.

nate-chandler committed:
[AddressLowering] Sink projections to uses.

Visit the tree of projections in post-order, sinking each to the lowest
position in the least common ancestor of all uses.

nate-chandler committed:
[AddressLowering] Project at storage.

The pass rewrites opaque values in reverse post order of their
definitions.  When an opaque value is rewritten, both its definition and
its uses are rewritten.  When a def or a use is rewritten, in order to
materialize an address with which to rewrite, projection instructions
are created.  Previously, these projections were created at the site of
the def or use.  Some of these projection instructions may be reused when
rewriting later opaque values.

As a result, it's possible to have two opaque values `A` and `B` (which
are rewritten in that order) such that rewriting a use of `A` which
occurs "after" the def (or a use) of `B` creates a projection `P` which
is then used by that "earlier" def (or use) of `B`:

    ```
    A =
    B =     // relies on P
    use B
    use A   // creates some projection P
    ```

When rewriting the def (or that use, respectively) of `B`, the
projection which was created for the use of `A` will be reused.  And
previously, the projection would be created at the use of A.  But that
projection instruction was "after" the place where it is used when
rewriting the def or use of `B`.  That's invalid!

To address this, rather than creating projections at the instruction
being rewritten, instead create them "as early as possible".  The
locations where they will be created are chosen so as to dominate all
uses.

nate-chandler committed:
[AddressLowering] Prevent existential multi-init.

nate-chandler commented:
@swift-ci please test

nate-chandler commented on lib/SILOptimizer/Mandatory/AddressLowering.cpp:
```diff
@@ -1228,6 +1253,25 @@ void OpaqueStorageAllocation::allocateValue(SILValue value) {
   createStackAllocationStorage(value);
 }
 
+SILValue AddressLoweringState::getEarliestStoragePoint(SILValue userValue,
```
Added this comment (edited) with an example.

nate-chandler commented on lib/SILOptimizer/Mandatory/AddressLowering.h:
```diff
@@ -163,6 +165,60 @@ class ValueStorageMap {
+  ProjectionIterator projection_end() const {
+    return ProjectionIterator(nullptr, *this);
+  }
+  iterator_range<ProjectionIterator> getProjections(SILValue value) const {
```
Fixed.

nate-chandler commented on lib/SILOptimizer/Mandatory/AddressLowering.cpp:
```diff
@@ -540,10 +540,31 @@ struct AddressLoweringState {
+  /// corresponding to userValue should be materialized.
+  ///
+  /// @returns nonnull instruction
+  SILInstruction *getProjectionInsertionPoint(SILValue userValue) const;
```
Fixed.

nate-chandler commented:
@atrick Added projection sinking using a leaf-first walk over the projections and a new function `sinkToUses`.

nate-chandler commented:
@atrick As you said, it is sufficient to iterate `reverse(valueStorageMap)` to get a post-order of the projections for sinking.  There is a bit of (irrelevant) churn of relative position of projection instructions before their uses https://github.com/apple/swift/compare/08db6f612af952d11f4fb5dd2f62fc7b3a55f3f1..533919168eeba309fd6f23cb56382ec088cc79c5 .

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@atrick Restored the `isStoreCopy` assert.

nate-chandler commented:
@swift-ci please test windows platform

atrick commented on lib/SILOptimizer/Mandatory/AddressLowering.cpp:
```diff
@@ -1290,6 +1306,9 @@ SILValue AddressLoweringState::getEarliestStoragePoint(SILValue userValue,
       return SILValue();
     }
+    if (auto *loi = getLatestOpeningInst(pair)) {
+      return cast<SingleValueInstruction>(loi);
```
Why do we only consider the first latest opening instruction? Can we way say something about the relationship between open archetypes based on the nature of use projections?

If we find a latest opening instruction, then why don't we also check dominance on the non-projection storage address the same way we do for the opening instruction?

Can we just lump all dependencies together and check dominance for each without caring about their relationship?

nate-chandler replied:
> Why do we only consider the first latest opening instruction?

This was a false assumption I was making.  I added a test case demonstrating that.  It failed when considering only the first such instruction.

> Can we way say something about the relationship between open archetypes based on the nature of use projections?

We can say that they all dominate or are dominated by one another thanks to them all dominating the largest aggregate.  But we can't say anything about their relative ordering based on when they start appearing in the projection tower.

> If we find a latest opening instruction, then why don't we also check dominance on the non-projection storage address the same way we do for the opening instruction?

Fixed.

> Can we just lump all dependencies together and check dominance for each without caring about their relationship?

Yep, done.

atrick commented on lib/SILOptimizer/Mandatory/AddressLowering.cpp:
```diff
@@ -1233,6 +1245,67 @@ void OpaqueStorageAllocation::allocateValue(SILValue value) {
+/// As a result of the above consideration, when this function is called for a
+/// phi, it returns a null value if the storage involves either sort of
+/// effectful initialization.
+SILValue AddressLoweringState::getEarliestStoragePoint(SILValue userValue,
```
Can we be explicit about the relationship between this function and  getProjectionInsertionPoint which is called later. Like, if this returns a valid address, the getProjectionInsertionPoint must be guaranteed to find a dominating address, so all dependencies discovered here need to be rediscovered there.

nate-chandler replied:
Fixed.

atrick commented on lib/SILOptimizer/Mandatory/AddressLowering.cpp:
```diff
@@ -1815,8 +1835,18 @@ SILInstruction *AddressMaterialization::getProjectionInsertionPoint(
+      // It's only necessary to consider the obstruction to the earliest
+      // projection because obstructions to any others must dominate it.
+      if (auto *loi = pass.getLatestOpeningInst(pair))
+        obstruction = loi->getNextInstruction();
```
Same question about obstructions: Why do we only consider the first latest opening instruction?
But this time we need to consider def projections. (Maybe that doesn't matter.)

nate-chandler replied:
Fixed to consider all obstructions.  Instead of collecting them into a list, it relies on the fact that the opening instructions all must either dominate or be dominated by one another.

nate-chandler committed:
[AddressLowering] NFC: Comment projection chains.

Describe what all chains look like by analyzing the four theoretically
possible kinds of links--only three of the four are actually possible.

nate-chandler committed:
[AddressLowering] Gardening: Detypo'd.

nate-chandler committed:
[AddressLowering] NFC: Moved assert above NPE.

Moved the assert that the value which opens an archetype exists before
the call to get its defining instruction which would be an NPE if the
value were in fact null.

nate-chandler committed:
[AddressLowering] Test: Added for arg behavior.

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
Please test with following pull request:
https://github.com/apple/swift-syntax/pull/1948

@swift-ci please test macos platform

atrick commented on lib/SILOptimizer/Mandatory/AddressLowering.cpp:
```diff
@@ -1248,36 +1302,60 @@ bool OpaqueStorageAllocation::findProjectionIntoUseImpl(
-      assert(isa<SILFunctionArgument>(baseStorage->storageAddress));
+    // Recurse through all storage projections to find the point where the
+    // storage has been allocated and any opening instructions involved in
+    // any of those projections' types.
```
This is all good.

But it may be more clear if we also outright say something like:

```
The base storage address and all of the opened types used by the projections must dominate 'incomingValues' because we need to materialize corresponding address projections at those points.
```

nate-chandler replied:
Done.

atrick reviewed (approved):
LGTM

nate-chandler commented:
@swift-ci please smoke test and merge
